### PR TITLE
Add Local implementation of abstract engine classes

### PR DESCRIPTION
**Pull request by dstrain115**
This PR creates partially implemented sub-classes that define
in-memory objects for engine functionality unlikely to be implemented
differently, such as reservations, labels, and descriptions:

AbstractLocalJob, AbstractLocalProgram,
AbstractLocalProcessor, AbstractLocalEngine

These classes follow the implementation of the production
Engine classes but use local memory to provide equivalent
functionality for testing, mocking, and local simulator use.

This is PR 2 of a three part series.

### CODE REVIEW

File: cirq-google/cirq_google/engine/abstract_local_engine.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_engine.py:48]: `Iteratibvely` -> `Iteratively`

File: cirq-google/cirq_google/engine/abstract_local_engine.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_engine.py:45]: Should `AbstractProgram` here be `AbstractLocalProgram`?

Similarly for other non-local types in this file.

File: cirq-google/cirq_google/engine/abstract_local_engine_test.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_engine_test.py:98]: Test getting an invalid processor?

File: cirq-google/cirq_google/engine/abstract_local_job.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_job.py:66]: What qualifies as an update? Should the `set_*` functions refresh this value?

File: cirq-google/cirq_google/engine/abstract_local_job_test.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_job_test.py:74]: `test_reps_and_Sweeps` -> `test_reps_and_sweeps`

File: cirq-google/cirq_google/engine/abstract_local_job_test.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_job_test.py:106]: Populate or remove empty tests.

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Reviewed `AbstractLocal(Engine|Job)`. Also see my comments from #4644, as many of the parent-class comments apply to the child class.

Will follow up with `AbstractLocal(Processor|Program)` reviews.

File: cirq-google/cirq_google/engine/abstract_local_processor.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_processor.py:262]: Missing return type annotation.

File: cirq-google/cirq_google/engine/abstract_local_processor.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_processor.py:292]: Missing return type annotation.

File: cirq-google/cirq_google/engine/abstract_local_processor.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_processor.py:324]: These are flipped - `reservation` should be excluded if its `end_time` is before `start_timestamp` or if its `start_time` is after `end_timestamp`.

File: cirq-google/cirq_google/engine/abstract_local_processor.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_processor.py:280]: Missing `Args` and `Raises` annotations.

File: cirq-google/cirq_google/engine/abstract_local_processor.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_processor.py:403]: `created after` -> `created before`

File: cirq-google/cirq_google/engine/abstract_local_processor_test.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_processor_test.py:105]: Recommend using a fake email here.

File: cirq-google/cirq_google/engine/abstract_local_processor_test.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_processor_test.py:128]: Similarly here - use fake emails if emails are required.

File: cirq-google/cirq_google/engine/abstract_local_program.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_program.py:30]: `structions` -> `structures`?

File: cirq-google/cirq_google/engine/abstract_local_program.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_program.py:34]: There's a disconnect here - line 33 seems incomplete, and 34 seems redundant.

File: cirq-google/cirq_google/engine/abstract_local_program.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/abstract_local_program.py:87]: `created after` -> `created before`

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Reviewed `AbstractLocal(Processor|Program)`.

File: cirq-google/cirq_google/engine/abstract_local_engine.py
Comment (dstrain115) [reply cirq-google/cirq_google/engine/abstract_local_engine.py:45]: I think this would violate the inheritance structure since this is defined in the interface as to return AbstractProgram.  AbstractLocalProgram is a subtype of AbstractProgram, so this is still true.

File: cirq-google/cirq_google/engine/abstract_local_job.py
Comment (dstrain115) [reply cirq-google/cirq_google/engine/abstract_local_job.py:66]: Good point, done.

File: cirq-google/cirq_google/engine/abstract_local_program.py
Comment (dstrain115) [reply cirq-google/cirq_google/engine/abstract_local_program.py:34]: Fixed.

**Review by dstrain115 — COMMENTED** [review]
Thanks for the thorough review.  I think I addressed all comments.

File: cirq-google/cirq_google/engine/abstract_local_processor.py
Comment (95-martin-orion) [reply cirq-google/cirq_google/engine/abstract_local_processor.py:122]: It sounds like this will work from your comment, but to confirm: will the following behave properly?
```
id1 = self._create_id()                          # id1 = '.../1/reservation'
id2 = self._create_id(id_type='something_else')  # id2 = '.../2/something_else'
id3 = self._create_id()                          # id3 = '.../3/reservation'
```
i.e. is it OK that the two 'reservation' IDs don't have consecutive numbers?

File: cirq-google/cirq_google/engine/abstract_local_processor.py
Comment (95-martin-orion) [reply cirq-google/cirq_google/engine/abstract_local_processor.py:280]: `end_time: New starting time` -> `end_time: New ending time`

**Review by 95-martin-orion — APPROVED** [review]
A couple small items left to address, then this is good to go once the part 1 PR is merged.

File: cirq-google/cirq_google/engine/abstract_local_processor.py
Comment (dstrain115) [reply cirq-google/cirq_google/engine/abstract_local_processor.py:122]: Yeah, this should be fine.  I think the reality is that the ids are UUIDs, not consecutive integers.  We just want them to be unique.  I think the one issue is that jobs should really be children of programs, but I am not going to worry about that right now.

File: cirq-google/cirq_google/engine/abstract_local_processor.py
Comment (dstrain115) [reply cirq-google/cirq_google/engine/abstract_local_processor.py:280]: Fixed.

**Comment by MichaelBroughton** [comment]
Looks like there are some conflicts in merging part 1 into this.

**Comment by dstrain115** [comment]
> Looks like there are some conflicts in merging part 1 into this.

Fixed.  The merge conflicts were due to removing the coverage:ignore statements, which was intentional, as they are now covered by the tests in this PR.

**Comment by mpharrigan** [comment]
At least `SimulatedLocalProcessor` does not implement the Cirq requirement of roundtrippable reprs

**Comment by mpharrigan** [comment]
or JSON serializability

**Pull request merged**
